### PR TITLE
Do not color file output with `--muted-logs`

### PR DESCRIPTION
**Pull request by IsaacPD**
Fixes: #4913 

**Description**

Wrap os.File in filelog.MutedLogger type that is checked on Color.Fprint to determine whether colors should be printed.

### **Example of change**

Running ```skaffold run --mute-logs=all``` in skaffold/examples/ruby
**Before**: 
```
...
Starting deploy...
- writing logs to /var/.../skaffold/deploy/2020-11-02_18-49-06.log
...
```
In /var/.../skaffold/deploy/2020-11-02_18-49-06.log
```
^[[34mStarting deploy...^[[0m
 - service/ruby configured
 - deployment.apps/ruby configured
```

The logs contain ANSI escape codes which do not represent the actual content intended to be logged.

**After**:
```
...
Starting deploy...
- writing logs to /var/.../skaffold/deploy/2020-11-02_18-53-51.log
...
```
In /var/.../skaffold/deploy/2020-11-02_18-53-51.log
```
Starting deploy...
 - service/ruby configured
 - deployment.apps/ruby configured
```

The only thing logged is the intended output.

### CODE REVIEW

**Comment by MarlonGamez** [comment]
Would this cause issues in a situation where output is written to a temporary buffer which is then dumped to the terminal?

**Comment by IsaacPD** [comment]
> @MarlonGamez Would this cause issues in a situation where output is written to a temporary buffer which is then dumped to the terminal?

As long as the write to the terminal is done with ```Color.Fprintf/Fprintln``` then there should be no issues with coloring to the terminal.

```go
var b bytes.Buffer
Blue.Fprintf(&b, "some blue text")
fmt.Fprintf(out, b)
``` 

In the case above the final output would not have any special coloring in "out" if that's what you're referring to.

**Review by tejal29 — APPROVED** [review]

**Comment by MarlonGamez** [comment]
When performing parallel builds, we're printing in a way that will result in a loss of colored logs. I believe each artifact being built basically has it's own buffer which is written to with a mix of colored and non-colored lines, and each buffer is then dumped to the terminal

**Comment by IsaacPD** [comment]
Yeah I would not like to lose colored output, so I'll see what approaches are open to preserve that.

**Comment by tejal29** [comment]
Thanks @MarlonGamez for going through the build in parallel flow. You are right, with the way to log for parallel builds, color output will be broken.

Wondering if we should create a `MuteLogger` object and use typecheck instead of the check `out != colorfulWriter `

**Comment by IsaacPD** [comment]
That approach seems alright with me as it tackles the underlying issue. I cannot really think of any other ways that ANSI output my showup but i'll rename this issue to reflect that it does not potentially cover every case.

File: pkg/skaffold/color/formatter.go [hunk 1]
```diff
@@ -113,10 +114,19 @@ func (c Color) Fprintln(out io.Writer, a ...interface{}) {
...
-	if c.color == nil {
+	if c.color == nil || isMuted(out) {
 		fmt.Fprintf(out, format, a...)
 		return
 	}
 
 	fmt.Fprint(out, c.color.Sprintf(format, a...))
 }
+
+func isMuted(out io.Writer) bool {
```
Comment (briandealwis) [inline pkg/skaffold/color/formatter.go:125]: Would it make sense to push this into ‘logfile‘?

**Review by briandealwis — COMMENTED** [review]
So thinking about it some more, this approach will continue to add colour sequences to other file output. I think we need to instead record the actual terminal io.Writer instance and compare the provided writer.

**Comment by briandealwis** [comment]
Ah, missed the other discussion on this PR. We may be hosed on parallel output since we're redirecting the output and so the underlying builders won't colour their output (e.g., maven and gradle don’t colour their output).

**Comment by briandealwis** [comment]
I'm thinking that we should create a coloured output: rather than our current practice:
```
func Xxxx(out io.Writer, ...) {
   color.Blue.Fprintf(out, "some blue text")
}
```

and trying to guess whether the output supports color, we instead create a Writer wrapper that carries whether it supports colour or not.
```
func main() {
   out := color.NewWriter(os.Stdout)
   Xxx(out)

   var b bytes.Buffer
   if color.isTerm(os.Stdout) { // or something like this
      b = color.NewByteBuffer(b)
   }
   Xxx(&b)
   b.WriteTo(os.Stdout)
}

func Xxxx(out io.Writer, ...) {
   color.Blue.Fprintf(out, "some blue text") // Fprintf could then see if the writer supports colour
}
```

Looks like we can force [Maven (for Jib) to colour output](https://mihai-nita.net/2020/08/23/force-maven-colors-in-eclipse-console/) with `-Dstyle.color=always -Djansi.passthrough=true`.  I'm sure there are equivalents for Gradle. 😱

**Comment by IsaacPD** [comment]
I think this approach makes sense. So in this instance, https://github.com/GoogleContainerTools/skaffold/blob/d667827c018c592abcfcdd5738f42eb5848adb48/pkg/skaffold/build/result.go#L56-L65

Something like this would be added to determine whether to color builds based on whether the final output (stdout or a log file) is also colorable.
```go
if color.isColorable(l.out) {
    w := color.NewWriter(w)
}
```

**Comment by briandealwis** [comment]
I just came across [github.com/creack/pty](https://github.com/creack/pty) which runs commands under a PTY so that commands are run under a pseudo terminal.  This seems like the right solution for parallel-mode build output.

**Comment by IsaacPD** [comment]
I was having some issues earlier trying to use these flags you gave @briandealwis "-Dstyle.color=always -Djansi.passthrough=true" since they weren't working, but it turns out the order of the flags is relevant and if these were at the end then they didn't do anything. It might not be necessary then to use https://github.com/creack/pty

**Comment by IsaacPD** [comment]
Closing in favor of #5014

**Pull request closed without merging**